Edits to rover dev docs (#1636)
apollographql/apollo-cli: docs/source/commands/dev.mdx

---
title: The Rover dev command
description: Run your supergraph in your local dev environment
---

> ⚠️ **Do not run `rover dev` in production!** This command is for local development _only_.

> A **supergraph** is an architecture consisting of multiple GraphQL APIs (**subgraphs**) and a **graph router** that runs in front of them:

Author: Just fiddling with the preamble while I'm here

```mermaid
graph LR;
  clients(Clients);
  subgraph "Supergraph";
  router(["Graph Router"]);
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  router --- serviceA & serviceB;
  end;
  clients -.- router;
  class clients secondary;
```

While you're making local changes to an individual subgraph, you can use the `rover dev` command to start a local router instance and test out the effects of those changes on your entire supergraph.

Whenever you add, modify, or remove a subgraph from your local supergraph, Rover automatically re-composes your individual subgraph schemas into a unified supergraph schema, which your local router automatically begins using.

## Starting a session

> **To use `rover dev`, you need at least one running GraphQL API (subgraph).** Rover can obtain a subgraph's schema via introspection (either standard or [federated introspection](https://www.apollographql.com/docs/federation/subgraph-spec#enhanced-introspection-with-query_service)), or you can provide a local schema file.

Reviewer: Should I have made that link relative rather than absolute? Like this:

```suggestion
**To use `rover dev`, you need at least one running GraphQL API (subgraph).** Rover can obtain a subgraph's schema via introspection (either standard or [federated introspection](/docs/federation/subgraph-spec#enhanced-introspection-with-query_service)), or you can provide a local schema file.
```

Author: ah yes! Except starting with `/federation` instead of `/docs/federation` because of gatsby weirdness. Will apply in followup commit

Here's an example `rover dev` command that points to a locally running subgraph and provides its schema via a local file:

```bash title="Example command"
rover dev --name products --schema ./products.graphql --url http://localhost:4000
```

> **You don't have to provide a _locally running_ subgraph!** You can point to _any_ GraphQL endpoint that your local environment can reach. You can even mix and match local and remote subgraphs, which is helpful for testing local changes against your staging subgraphs.

**When you start a `rover dev` session:**

1. Rover obtains the subgraph schema you provide (via either introspection or file path).
2. Rover composes a [supergraph schema](/federation/federated-types/overview/#supergraph-schema) from the subgraph schema.
3. Rover starts a locally running [router](/router) instance and provides it the supergraph schema.
4. Rover starts watching the provided subgraph schema for changes (via either introspection or file), and it recomposes the supergraph schema whenever it detects a change. This automatically reloads the router.

After you start a session, you can then [add other subgraphs to it.](#adding-a-subgraph-to-a-session)

> ### Starting a session with multiple subgraphs

Author: Because this is unique to _starting_ the session, I made it a subitem of of the Starting section

For example, this `supergraph.yaml` file provides the necessary details for two subgraphs:

```yaml title="supergraph.yaml"
federation_version: =2.4.7
subgraphs:
  products:
    routing_url: http://localhost:4000
    schema:
      file: ./products.graphql # Schema provided via file
  reviews:
    routing_url: http://localhost:4001
    schema:
      subgraph_url: http://localhost:4001 # Schema provided via introspection
```

You provide this file to `rover dev` like so:

```bash
rover dev --supergraph-config supergraph.yaml
```

If you do, a `rover dev` session starts with _one_ of the subgraphs listed, then adds the remaining subgraphs one at a time (order is undefined). Because of this, you might observe composition errors during intermediate steps.

> Providing a `supergraph.yaml` file also enables you to take advantage of [other config options](./supergraphs#yaml-configuration-file), such as `introspection_headers`.

After you start your initial `rover dev` session with a config file, you can still [add other subgraphs individually](#adding-a-subgraph-to-a-session). However, you _can't_ provide another config file.

## Adding a subgraph to a session

> After you start a `rover dev` session with one subgraph, you can then add _other_ subgraphs to that same session.

Reviewer: ```suggestion
After you start an initial `rover dev` session, you can then add _other_ subgraphs to that same session.
```

> To add a subgraph, open a new terminal window and run `rover dev` _again_, this time providing the details of the subgraph to add. For example, this command adds a `reviews` subgraph to the session started with a `products` subgraph above:
> 
> ```bash
> rover dev --name reviews --url http://localhost:4001

Reviewer: Since `reviews` in this scenario would conflict with `reviews` added via config
```suggestion
To add a subgraph, open a new terminal window and run `rover dev` _again_, this time providing the details of the subgraph to add. For example, this command adds a `users` subgraph:

```bash
rover dev --name users --url http://localhost:4002
```

```

Rover detects your _existing_ session and attaches this new process to it.

> When you add a new subgraph to a session, Rover recomposes the supergraph schema and updates the router, so you can start querying _multiple_ subgraphs via the single router endpoint.

Reviewer: You could already query multiple via config above—though maybe it's clear that this just another way to do that?

Author: rephrased slightly 👍

> ⚠️ **Rover uses the port of the running router to identify an existing session.** If you specify a custom port via `--supergraph-port` or `--router-config`, make sure to specify the _same_ port for all `rover dev` processes that you want to attach to the same session.

> ## Stopping a session

Reviewer: ```suggestion
## Stopping all sessions
```
